File: llg.c
#include "clib.h"

void llg_rhs(double *dm_dt, double *m, double *h, double gamma, double alpha,
		double mu_s, int nxyz, double c) {

	int i, j, k;

	double mth0, mth1, mth2;
	double coeff = -gamma / (1 + alpha * alpha) / mu_s;
	double mm, relax;

	for (i = 0; i < nxyz; i++) {
		j = i + nxyz;
		k = j + nxyz;

		mth0 = coeff * (m[j] * h[k] - m[k] * h[j]);
		mth1 = coeff * (m[k] * h[i] - m[i] * h[k]);
		mth2 = coeff * (m[i] * h[j] - m[j] * h[i]);

		dm_dt[i] = mth0 + alpha * (m[j] * mth2 - m[k] * mth1);
		dm_dt[j] = mth1 + alpha * (m[k] * mth0 - m[i] * mth2);
		dm_dt[k] = mth2 + alpha * (m[i] * mth1 - m[j] * mth0);

		mm = m[i] * m[i] + m[j] * m[j] + m[k] * m[k];
		relax = c * (1 - mm);
		dm_dt[i] += relax * m[i];
		dm_dt[j] += relax * m[j];
		dm_dt[k] += relax * m[k];

	}

}

File: fidimag/atomistic/llg.py
from __future__ import division
from __future__ import print_function

import fidimag.extensions.clib as clib

from .atomistic_driver import AtomisticDriver


class LLG(AtomisticDriver):

    """

    This class is the driver to solve the Landau Lifshitz Gilbert equation
    which has the form:


          ds        -gamma
         ---- =    --------  ( s X H_eff  + a * s X ( s X H_eff ) )
          dt             2
                  ( 1 + a  )

    by using the Sundials library with CVODE.

    This class inherits common methods to evolve the system using CVODE, from
    the micro_driver.MicroDriver class. Arrays with the system information
    are taken as references from the main micromagnetic Simulation class

    """

    def __init__(self, mesh, spin, mu_s, mu_s_inv, field, pins,
                 interactions,
                 name,
                 data_saver,
                 use_jac
                 ):

        # Inherit from the driver class
        super(LLG, self).__init__(mesh, spin, mu_s, mu_s_inv, field,
                                  pins, interactions, name,
                                  data_saver,
                                  use_jac
                                  )

    def sundials_rhs(self, t, y, ydot):

        self.t = t

        # already synchronized when call this funciton
        # self.spin[:]=y[:]

        self.compute_effective_field(t)

        clib.compute_llg_rhs(ydot,
                             self.spin,
                             self.field,
                             self._alpha,
                             self._pins,
                             self.gamma,
                             self.n,
                             self.do_precession,
                             self.default_c)

        #ydot[:] = self.dm_dt[:]

        return 0

if __name__ == '__main__':
    pass
